Route relation labeling (again)

tyrasd opened the pull request:
A slightly different approach for #8759 (and a bit beyond) to fix the issue with overly long NSI preset names for PT route relations. Supersedes  #10583[^1].

* Uses the regular non-NSI base preset for the labeling in the relation membership (and relation members) editors
* Shows the icon of the NSI preset next to the base preset name
* Shows the route's `ref` tag in a (colored[^2]) "shield" 

Some before/after examples:

<img src="https://github.com/user-attachments/assets/f38e6f09-ffef-4291-94bf-434795787017" width="300"> - <img src="https://github.com/user-attachments/assets/b83cb073-c398-48ca-9ef5-701c058e8648" width="300">

<hr>

<img src="https://github.com/user-attachments/assets/948542de-2bbf-4992-b08c-b0dcb3f10a2f" width="300"> - <img src="https://github.com/user-attachments/assets/670cbb0e-d809-480a-9926-42c07396feb5" width="300">

<hr>

<img src="https://github.com/user-attachments/assets/92768e35-d8af-4de2-a492-3acf55ece024" width="300"> - <img src="https://github.com/user-attachments/assets/ea463e9e-e81f-48e2-a840-dadf58bfcce1" width="300">

<hr>

<img src="https://github.com/user-attachments/assets/a0c69481-5c32-43e5-a51f-be7431f50b41" width="300"> - <img src="https://github.com/user-attachments/assets/ec79741c-8afa-44d3-b295-20596e30bff5" width="300">

<hr>

[^1]: Different to the approach in #10583, this is not relying on implicit assumptions about NSI's naming scheme and does not interfere with (translated) names of regular presets.
[^2]: If applicable, see also #9424

tyrasd committed:
don't use NSI preset names for labels in raw member(ship) editor

fixes #8759, closes #10583

tyrasd committed:
show full entity name as hover-text in search results

in case entries with long names (e.g. public transport route relations that have a NSI entry) can still be disambiguated from each other

tyrasd committed:
show (route) ref in a (colored) shield in relation editors + search results

1ec5 commented on modules/ui/feature_list.js:
```diff
@@ -332,12 +333,26 @@ export function uiFeatureList(context) {
+                const hasColor = d.entity.tags.colour && isColourValid(d.entity.tags.colour);
+                const hasRef = d.entity.tags.ref;
+                if (hasColor || hasRef) {
+                    const color = isColourValid(d.entity.tags.colour) ? d.entity.tags.colour : '#555';
```
Many recreational routes are associated with multiple colors, which could be tagged as a semicolon-delimited list in `colour=*` (regardless of any `osmc:symbol=*`).

tyrasd replied:
Yeah, I was pondering about `osmc:symbol` for hiking routes (but only for a future iteration on top of this feature). Multi-value `colour` tags seem to exist ([example](https://www.openstreetmap.org/relation/7840712)), but are not very frequent outside of the Netherlands:

[!<img src="https://github.com/user-attachments/assets/1de64420-d760-4172-af63-123a3db4efe1" width="360">](https://overpass-turbo.eu/s/24ld)

1ec5 commented on modules/ui/feature_list.js:
```diff
@@ -332,12 +333,26 @@ export function uiFeatureList(context) {
 
+            label.each(function(d) {
+                if (d.entity?.type !== 'relation') return;
+                const hasColor = d.entity.tags.colour && isColourValid(d.entity.tags.colour);
```
Recently on the forum we were [discussing alternative keys](https://community.openstreetmap.org/t/define-check-and-automated-correct-string-values-for-the-tag-colour-for-buildings/129467/10) that often associate routes with colors. `colour=*` is usually appropriate when a route is associated with a color. However, `ref:colour=*` is used in some places, particularly Spain, where they don’t attempt to uniquely color-code the routes in a network, but the color is still very important for recognizability. If iD uses the color as a background color for something that looks like a route marker, then this would be a good opportunity to use this key. Unfortunately, id-tagging-schema doesn’t have a `ref:colour=*` field yet. It’s a touchy subject, because people who don’t understand the Spanish system or others like it tend to rush to judgment about how they’re doing it wrong.

matkoniecz replied:
> It’s a touchy subject, because people who don’t understand the Spanish system or others like it tend to rush to judgment about how they’re doing it wrong.

I have not rushed to yell "you are doing it wrong way" but I definitely had such thought. Is there write-up somewhere which explains why `ref:colour` rather than `colour` makes sense?

(if it makes sense it would make sense to support it, including iD presets)

1ec5 replied:
[The `ref:colour=*` documentation](https://wiki.openstreetmap.org/wiki/Key:ref:colour#Disambiguation) alludes to the situation. It’s primarily important in Spain, where colors on route shields apparently have political overtones. I’ve had someone explain the considerations there to me before, but it goes over my head. From my perspective as an OSM Americana developer, some of the distinctions seem like they could be represented by more specific `network=*` values instead, but other route colors seem to have no rhyme or reason to them, which would justify the presentational keys. Might be worth asking on the forum for a fuller explanation.

matkoniecz replied:
I am mostly confused why not use regular `colour` - I ma not asking why it gets tagged but why use different key than normal?

1ec5 replied:
[This highway route](https://www.openstreetmap.org/relation/14047150) is `colour=red` because it’s known as the Red Route; one would identify it by color alone. [This subway route](https://www.openstreetmap.org/relation/2851728) is tagged (more or less) `colour=green` because it’s known as the Green Line and the branding is green through and through.

By contrast, [AP-1](https://www.openstreetmap.org/relation/404438) has (more or less) `ref:colour=blue` because the _route markers_ are blue. That’s the only aspect of the route that’s blue. Normally we don’t tag this kind of detail, because all the routes in a given network share the same route marker style, but Spanish routes are more idiosyncratic.

tyrasd committed:
update test specs with new function parameter

tyrasd committed:
use ref:colour for relation colouring if present

tyrasd committed:
render multiple "ref shields" for semicolon separated ref values

tyrasd committed:
only render ref shields on relations in raw member editor

and also render multiple ref shields in search results (if applicable)
and drop some now unused imports

tyrasd committed:
Merge branch 'develop' into route-relation-labels

tyrasd committed:
fix copy/paste mistake

matkoniecz commented on css/80_app.css:
```diff
@@ -2894,21 +2881,35 @@ img.tag-reference-wiki-image {
+    margin-bottom: -4.5px;
+}
+.section-raw-membership-editor .member-row span.member-entity-ref-color,
+.section-raw-member-editor .member-row span.member-entity-ref-color,
```
section-raw-member-editor?

where it has effects?

1ec5 replied:
The member editor is what appears at the bottom of the inspector when a relation is selected. The membership editor is what appears at the bottom of the inspector when a relation’s member is selected.

matkoniecz replied:
so `raw-membership-editor` is for membership editor?

There is no special raw membership editor?

1ec5 replied:
Not sure what you’re asking, but the member and membership editors appear in the same location in the UI, in mutually exclusive situations. It does make sense to style them both consistently, as `ref=*` and `network=*` can both appear on relations and relation members, depending on the feature type.

matkoniecz replied:
> but the member and membership editors appear in the same location in the UI, in mutually exclusive situations

it does not seem mutually exclusive?

what about relations that are having members (so member editor is appearing) and are also part of other relations (route master one for example) so membership editor is appearing.

In fact membership editor will appears always, I think to enable adding relation into another relation as member.

(I am not nitpicking for sake of nitpicking, I am trying to confirm that I understood what you wanted to express)

1ec5 replied:
> what about relations that are having members (so member editor is appearing) and are also part of other relations (route master one for example) so membership editor is appearing.

Yes, you would see both in that case.

matkoniecz commented on modules/osm/tags.js:
```diff
@@ -339,6 +341,22 @@ export function isColourValid(value) {
+    else if (tags.colour)         color = tags.colour;
+    const isValid = isColorValid(color);
+    if (!isValid)                 color = fallback;
+    if (tags['ref:colour_tx'])    textColor = tags['ref:colour_tx'];
```
note that relation with `colour=red` `ref:colour_tx=red` will result in unreadable text

is it OSM data bug?

I see no doc at https://taginfo.openstreetmap.org/keys/ref%3Acolour_tx#wiki

I see mention at https://wiki.openstreetmap.org/wiki/Key:ref:colour which suggests it would be bug, so I guess that using it as is is fine?

matkoniecz replied:
what about

```
ref=REF
colour=#00ffff
ref:colour=orange
ref:colour_bg=red
ref:colour_tx=orange
route=bicycle
type=route
```

should it appear orange on orange?

See also https://www.openstreetmap.org/relation/15627027 and https://www.openstreetmap.org/note/5180615

tyrasd replied:
> is it OSM data bug?

yes, I would say so

matkoniecz commented:
I asked on https://t.me/OSMes/175014 for feedback

Not sure is it necessary to supporti `ref:colour_bg` somehow.

Robot8A commented:
It's something the Spanish community came with for road shields, but I see how it could be used beyond it too.

For example in Zaragoza, Spain, each public transport line has a specific visual representation:
https://detalier.com/portfolio/red-de-autobuses-de-zaragoza/ (this is the website of the designer studio who did the visual representation, as the official website of the bus company is inaccesible from a non-Spanish IP, but it can be seen there too)

Day lines are shown with a given background colour, with white text on stops. Night lines instead are dark blue background and a given text colour.

The question is whether that is interesting to show in iD or not.

And also probably how if we have so much complexity with osmc:symbol, why couldn't we have it as well for ref colour.

tyrasd commented:
> Not sure is it necessary to supporti `ref:colour_bg` somehow.

I would say not, because [`ref:colour_bg`](https://wiki.openstreetmap.org/wiki/Key:ref:colour) seems to be intended for the _background colour of the road sign_, which we don't care about here.

tyrasd committed:
Merge branch 'develop' into route-relation-labels

tyrasd committed:
add to changelog

hlfan commented:
> fix the issue with overly long NSI preset names for PT route relations

⚖️

<img src="https://github.com/user-attachments/assets/1081667c-8aba-440a-9c8a-c85966cf7dd9" />

matkoniecz commented:
it is kind of fitting for trains to have train-shaped icons...

well, I guess that solution would be to have limit on how long icon is allowed to be? Have you created already an issue for that?

1ec5 commented:
Long wordmark icons are precisely why name-suggestion-index has historically preferred Twitter and Facebook icons (which have to be square) over Wikimedia Commons logos. Does ÖBB NightJet have a [small logo or icon <small>(P8972)</small>](https://www.wikidata.org/wiki/Property:P8972) that could be added to [this Wikidata item](https://www.wikidata.org/wiki/Q28127371)? I believe NSI already prefers small logos over full logos.

tordans commented:
Site note: we are putting too much stuff onto this one line...

Robot8A commented:
> Long wordmark icons are precisely why name-suggestion-index has historically preferred Twitter and Facebook icons (which have to be square) over Wikimedia Commons logos. Does ÖBB NightJet have a [small logo or icon (P8972)](https://www.wikidata.org/wiki/Property:P8972) that could be added to [this Wikidata item](https://www.wikidata.org/wiki/Q28127371)? I believe NSI already prefers small logos over full logos.

Probably a version with just the letters NJ, which are used when abbreviating the train service (see https://iechub.rfi.it/ArriviPartenze/ArrivalsDepartures/Monitor?Arrivals=False&Search=tarvisio&PlaceId=3443 during night time as an example)

Edit: I wrote my message without realizing @hlfan actually already solved this https://www.wikidata.org/w/index.php?title=Q28127371&diff=prev&oldid=2469124147
